### PR TITLE
Introduce instanceHasRefs flag

### PR DESCRIPTION
**Pull request by tajila**
Introduce instanceHasRefs flag

There is an issue with instance description bits with respect to
flattened types. See https://github.com/eclipse/openj9/issues/10070.
Currently, there is a mistach between the layout of a type on its own,
and when it is flattened. This is due to the pre-padding required for
standalone valuetypes on compressedrefs mode. The `backfill` has been
used to track pre-padding. This is an error prone approach and has cause
bugs in the past.

The only cases where pre-padding is needed is on compressedRefs where
there is at least one double-slot sized field. A better solution (to
backfill approach) is, instead of pre-padding, insert a single slot
field in the spot where the padding would be. This this means that if a
type (in compressedrefs mode) has at least one single slot or ref field
the pre-padding is not needed. This leaves fields with only double slot
fields. In these cases, we do not need to track the padding because
double-slot fields cannot be references, so there is no mismatch with
the instance description bits between the standalone layout and the
flattened layout.

This PR is the first step in solving this problem. This PR introduces a
flag on the J9Class that indicates whether the type has instances field
which are references. In these cases the barrier API for copying
instance fields doesn't need to scan the instance description bits to
apply barrier loads/stores. It can simply use a slot sized load/store
for the operation.

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
@gacholio @amicic please review these changes

File: runtime/oti/j9nonbuilder.h
Comment (gacholio) [inline runtime/oti/j9nonbuilder.h:84]: I think this would be better with the `ClassClass`

File: runtime/vm/createramclass.cpp
Comment (gacholio) [inline runtime/vm/createramclass.cpp:1878]: Should this not set the alignment constraint flag from above?

File: runtime/vm/description.c
Comment (gacholio) [inline runtime/vm/description.c:297]: Comment.

File: runtime/vm/description.c
Comment (gacholio) [inline runtime/vm/description.c:306]: Comment.

File: runtime/vm/description.c
Comment (gacholio) [inline runtime/vm/description.c:304]: Should be `if defined`

File: runtime/gc_include/ObjectAccessBarrierAPI.hpp
Comment (amicic) [inline runtime/gc_include/ObjectAccessBarrierAPI.hpp:441]: This check is not needed here, and in fact is more future proof to not have it. Also we may for example have a (non-GC functional) barrier that profiles accesses for primitive types.

File: runtime/gc_include/ObjectAccessBarrierAPI.hpp
Comment (amicic) [inline runtime/gc_include/ObjectAccessBarrierAPI.hpp:448]: @RSalman This needs to be fixed eventually. This path should not be allowed if WB type is SATB.

File: runtime/gc_include/ObjectAccessBarrierAPI.hpp
Comment (amicic) [inline runtime/gc_include/ObjectAccessBarrierAPI.hpp:445]: I'm on a fence even if hasReferences is needed here, but let's leave it.
However, the check was not quite complete. It was missing  '.... || (j9gc_modron_wrtbar_always == _writeBarrierType)' (see how we treat stores)
Please, fix it while here.

File: runtime/gc_include/ObjectAccessBarrierAPI.hpp
Comment (RSalman) [reply runtime/gc_include/ObjectAccessBarrierAPI.hpp:448]: Noted to address this as part of SATB work

**Comment by tajila** [comment]
@amicic @gacholio Please take another look at these changes

File: runtime/gc_base/ObjectAccessBarrier.cpp
Comment (gacholio) [inline runtime/gc_base/ObjectAccessBarrier.cpp:1426]: Space after if.

File: runtime/gc_base/ObjectAccessBarrier.cpp
Comment (gacholio) [inline runtime/gc_base/ObjectAccessBarrier.cpp:1440]: Space after if

File: runtime/gc_base/ObjectAccessBarrier.cpp
Comment (gacholio) [inline runtime/gc_base/ObjectAccessBarrier.cpp:1513]: Space after if

File: runtime/gc_base/ObjectAccessBarrier.cpp
Comment (gacholio) [reply runtime/gc_base/ObjectAccessBarrier.cpp:1426]: I'll withdraw these as they appeared in the original code.

File: runtime/gc_base/ObjectAccessBarrier.cpp
Comment (gacholio) [inline runtime/gc_base/ObjectAccessBarrier.cpp:1554]: There's no need for the zero check here (and elsewhere this pattern appears).

**Review by gacholio — APPROVED** [review]
Approving assuming various nitpicks are addressed. Will leave this to @amicic to do the final merge.

File: runtime/gc_base/ObjectAccessBarrier.cpp
Comment (tajila) [reply runtime/gc_base/ObjectAccessBarrier.cpp:1554]: We are currently using this to indicate the presence of pre-padding in valuetypes. If this check is skipped then the instance description bits will be misalligned in the flattened representation (since padding is removed in these cases). 

I plan to address this in the next PR.

**Comment by tajila** [comment]
@amicic please take a look at these changes

**Comment by amicic** [comment]
didn't you need to adjust MM_ObjectAccessBarrierAPI::structuralFlattenedCompareObjects() similarly to copyObjectFields()?

**Comment by tajila** [comment]
@amicic 
> didn't you need to adjust MM_ObjectAccessBarrierAPI::structuralFlattenedCompareObjects() similarly to copyObjectFields()?

Yes, I have made the required changes there as well

**Comment by tajila** [comment]
@amicic Ive updated the PR

File: runtime/gc_include/ObjectAccessBarrierAPI.hpp
Comment (amicic) [inline runtime/gc_include/ObjectAccessBarrierAPI.hpp:415]: expand this with '|| (j9gc_modron_wrtbar_always == _writeBarrierType))' similarly to the other site

Comment (tajila) [reply runtime/gc_include/ObjectAccessBarrierAPI.hpp:415]: even though this is a readonly operation?

File: runtime/gc_base/ObjectAccessBarrier.cpp
Comment (amicic) [inline runtime/gc_base/ObjectAccessBarrier.cpp:1433]: this is more a question for my benefit: if these are not object pointers, what are they (and still care if it's compressed or not)?

Comment (tajila) [reply runtime/gc_base/ObjectAccessBarrier.cpp:1433]: These are primitive fields. We still need slot sized loads because each iteration needs to match with the instance description bits. Whether we are compressed or not will determine how bytes each description bit represents.

File: runtime/gc_include/ObjectAccessBarrierAPI.hpp
Comment (amicic) [reply runtime/gc_include/ObjectAccessBarrierAPI.hpp:415]: ah, good point. then, no.
the other one was a mix of both reads and writes, so I was hasty to assume this was the same case.

**Review by amicic — APPROVED** [review]

**Comment by amicic** [comment]
Jenkins test sanity win,aix jdk11

**Pull request merged**
